# Determinación de los valores de b y w de una línea recta

## Objetivo

El objetivo es encontrar los mejores valores de `b` y `w` para la ecuación:

y = b + wx

Se utilizará el dataset Salary, donde:

- `X` = YearsExperience (años de experiencia)
- `Y` = Salary (salario)

Se utilizarán los 35 registros del dataset para calcular los valores de `b` y `w`.

In [1]:
import numpy as np
data = np.loadtxt(
    'salary.txt',
    delimiter=',',
    skiprows=1
)
print(data)

[[1.10000e+00 3.93430e+04]
 [1.30000e+00 4.62050e+04]
 [1.50000e+00 3.77310e+04]
 [2.00000e+00 4.35250e+04]
 [2.20000e+00 3.98910e+04]
 [2.90000e+00 5.66420e+04]
 [3.00000e+00 6.01500e+04]
 [3.20000e+00 5.44450e+04]
 [3.20000e+00 6.44450e+04]
 [3.70000e+00 5.71890e+04]
 [3.90000e+00 6.32180e+04]
 [4.00000e+00 5.57940e+04]
 [4.00000e+00 5.69570e+04]
 [4.10000e+00 5.70810e+04]
 [4.50000e+00 6.11110e+04]
 [4.90000e+00 6.79380e+04]
 [5.10000e+00 6.60290e+04]
 [5.30000e+00 8.30880e+04]
 [5.90000e+00 8.13630e+04]
 [6.00000e+00 9.39400e+04]
 [6.80000e+00 9.17380e+04]
 [7.10000e+00 9.82730e+04]
 [7.90000e+00 1.01302e+05]
 [8.20000e+00 1.13812e+05]
 [8.70000e+00 1.09431e+05]
 [9.00000e+00 1.05582e+05]
 [9.50000e+00 1.16969e+05]
 [9.60000e+00 1.12635e+05]
 [1.03000e+01 1.22391e+05]
 [1.05000e+01 1.21872e+05]
 [1.12000e+01 1.27345e+05]
 [1.15000e+01 1.26756e+05]
 [1.23000e+01 1.28765e+05]
 [1.29000e+01 1.35675e+05]
 [1.35000e+01 1.39465e+05]]


La primera columna corresponde a YearsExperience, por lo que se utiliza como X. La segunda columna corresponde a Salary, por lo que se utiliza como Y.

Se obtiene la cantidad de registros del dataset. En este caso se utilizan 35 datos.

In [2]:
X = data[:, 0]
y=data[:, 1]
n=len(X)
print("X:", X)
print("Y:", y)
print(n)


X: [ 1.1  1.3  1.5  2.   2.2  2.9  3.   3.2  3.2  3.7  3.9  4.   4.   4.1
  4.5  4.9  5.1  5.3  5.9  6.   6.8  7.1  7.9  8.2  8.7  9.   9.5  9.6
 10.3 10.5 11.2 11.5 12.3 12.9 13.5]
Y: [ 39343.  46205.  37731.  43525.  39891.  56642.  60150.  54445.  64445.
  57189.  63218.  55794.  56957.  57081.  61111.  67938.  66029.  83088.
  81363.  93940.  91738.  98273. 101302. 113812. 109431. 105582. 116969.
 112635. 122391. 121872. 127345. 126756. 128765. 135675. 139465.]
35


Se crean variables que comenzarán en cero y servirán para acumular las sumatorias necesarias para calcular w y b

In [3]:
suma_x = 0
suma_y = 0
suma_xy = 0
suma_x2 = 0

Se utiliza un ciclo for para recorrer los 35 registros y calcular las sumatorias de X, Y, X*Y y X²

In [4]:
for i in range(n):

    suma_x = suma_x + X[i]

    suma_y = suma_y + y[i]

    suma_xy = suma_xy + (X[i] * y[i])

    suma_x2 = suma_x2 + (X[i] * X[i])

Se calcula w, que representa la pendiente de la línea. El resultado obtenido fue aproximadamente:

In [5]:
numerador = (n * suma_xy) - (suma_x * suma_y)

denominador = (n * suma_x2) - (suma_x * suma_x)

w = numerador / denominador

print("Valor de w:", w)

Valor de w: 8731.941047062492


Esto significa que el salario aumenta aproximadamente 8731.94 por cada año adicional de experiencia, según el modelo.

Se calcula b, que representa el punto donde la línea corta el eje Y

In [6]:
promedio_x = suma_x / n

promedio_y = suma_y / n

b = promedio_y - (w * promedio_x)

print("Valor de b:", b)

Valor de b: 28859.526194531485


Se utiliza nuevamente un for para calcular el salario estimado de cada uno de los 35 registros utilizando la ecuación obtenida.

In [7]:
print("Predicciones:")
print()

for i in range(n):

    prediccion = b + (w * X[i])

    print(
        "x =", X[i],
        "Real =", y[i],
        "Predicción =", round(prediccion, 2)
    )

Predicciones:

x = 1.1 Real = 39343.0 Predicción = 38464.66
x = 1.3 Real = 46205.0 Predicción = 40211.05
x = 1.5 Real = 37731.0 Predicción = 41957.44
x = 2.0 Real = 43525.0 Predicción = 46323.41
x = 2.2 Real = 39891.0 Predicción = 48069.8
x = 2.9 Real = 56642.0 Predicción = 54182.16
x = 3.0 Real = 60150.0 Predicción = 55055.35
x = 3.2 Real = 54445.0 Predicción = 56801.74
x = 3.2 Real = 64445.0 Predicción = 56801.74
x = 3.7 Real = 57189.0 Predicción = 61167.71
x = 3.9 Real = 63218.0 Predicción = 62914.1
x = 4.0 Real = 55794.0 Predicción = 63787.29
x = 4.0 Real = 56957.0 Predicción = 63787.29
x = 4.1 Real = 57081.0 Predicción = 64660.48
x = 4.5 Real = 61111.0 Predicción = 68153.26
x = 4.9 Real = 67938.0 Predicción = 71646.04
x = 5.1 Real = 66029.0 Predicción = 73392.43
x = 5.3 Real = 83088.0 Predicción = 75138.81
x = 5.9 Real = 81363.0 Predicción = 80377.98
x = 6.0 Real = 93940.0 Predicción = 81251.17
x = 6.8 Real = 91738.0 Predicción = 88236.73
x = 7.1 Real = 98273.0 Predicción = 90856.

Se crean 10 nuevos valores de años de experiencia

In [8]:
nuevos_x = [
    1.2,
    1.7,
    2.5,
    3.5,
    4.3,
    5.5,
    6.5,
    7.5,
    8.5,
    10.0
]

Se utiliza la ecuación encontrada para estimar el salario correspondiente a cada uno de los 10 nuevos valores de experiencia.

In [9]:
for i in range(len(nuevos_x)):

    nuevo_y = b + (w * nuevos_x[i])

    print(
        "Años de experiencia:", nuevos_x[i],
        "Salario estimado:", round(nuevo_y, 2)
    )

Años de experiencia: 1.2 Salario estimado: 39337.86
Años de experiencia: 1.7 Salario estimado: 43703.83
Años de experiencia: 2.5 Salario estimado: 50689.38
Años de experiencia: 3.5 Salario estimado: 59421.32
Años de experiencia: 4.3 Salario estimado: 66406.87
Años de experiencia: 5.5 Salario estimado: 76885.2
Años de experiencia: 6.5 Salario estimado: 85617.14
Años de experiencia: 7.5 Salario estimado: 94349.08
Años de experiencia: 8.5 Salario estimado: 103081.03
Años de experiencia: 10.0 Salario estimado: 116178.94


In [10]:
if w > 0:

    print("La relación entre experiencia y salario es positiva.")

elif w < 0:

    print("La relación entre experiencia y salario es negativa.")

else:

    print("No existe una relación lineal.")

La relación entre experiencia y salario es positiva.


Se utiliza una estructura decisiva if, elif y else para determinar el tipo de relación entre las variables.

Como w es positivo, el programa indica que existe una relación positiva: a mayor experiencia, mayor salario estimado.

# Conclusiones

Se utilizó el dataset Salary, compuesto por 35 registros de años de experiencia y salario.

Mediante un algoritmo implementado manualmente se calcularon los valores de `b` y `w`, sin utilizar funciones automáticas de regresión.

Los resultados obtenidos fueron:

- `w = 8731.94`
- `b = 28859.53`

Por lo tanto, la ecuación obtenida fue:

y = 28859.53 + 8731.94x

El valor positivo de `w` indica que existe una relación positiva entre los años de experiencia y el salario. Es decir, según los datos utilizados, al aumentar los años de experiencia también aumenta el salario estimado.

Finalmente, se realizaron 10 inferencias con nuevos valores de `X`, obteniendo los salarios estimados mediante la ecuación encontrada.

El algoritmo utilizó estructuras repetitivas `for` para recorrer los datos y una estructura decisiva `if/elif/else` para interpretar la relación entre las variables.